File: DS/UnionFind/connectingGraph.py
class ConnectingGraph:
    """
    @param: n: An integer
    """
    def __init__(self, n):
        # do intialization if necessary
        self.father = { i : i for i in range(1, n + 1)}
        self.count = n 
    """
    @param: a: An integer
    @param: b: An integer
    @return: nothing
    """
    def connect(self, a, b):
        # write your code here
        
        if a not in self.father or b  not in self.father:
            
            return 
        
        self.union(a, b)

    """
    @param: a: An integer
    @param: b: An integer
    @return: A boolean
    """
    def query(self, a, b):
        # write your code here
        
        if a not in self.father or b not in self.father:
            
            return False 
        
        root_a = self.find(a)
        root_b = self.find(b)
        
        return root_a == root_b
        
    def union(self, a, b):
        
        root_a = self.find(a)
        root_b = self.find(b)
        
        if root_a == root_b:
            
            return 
        
        self.father[root_b] = root_a
        
        self.count -= 1 
        
        
    def find(self, point):
        
        path = [] 
        
        while point != self.father[point]:
            
            path.append(point)
            
            point = self.father[point]
            
        for p in path:
            
            self.father[p] = point 
            
        return point 

File: interviews/ForUsAll/Java/ConnectingGraph.java
import java.util.ArrayList;
import java.util.HashMap;

public class ConnectingGraph {
    
    HashMap<Integer, Integer> father;
    int count;
    /*
    * @param n: An integer
    */public ConnectingGraph(int n) {
        // do intialization if necessary
        this.father = new HashMap<Integer, Integer>();
        this.count = n; 
        
        for (int i = 1; i < n + 1; i++) {
            this.father.put(i, i);
        }
    }

    /*
     * @param a: An integer
     * @param b: An integer
     * @return: nothing
     */
    public void connect(int a, int b) {
        // write your code here
        
        int root_a = this.find(a);
        int root_b = this.find(b); 
        
        if (root_a == root_b) return;
        
        this.father.put(root_b, root_a); 
        this.count -= 1;
    }
    
    public int find(int point) {
        
        ArrayList<Integer> path = new ArrayList<>();
        
        while (point != this.father.get(point)) {
            
            path.add(point);
            point = this.father.get(point);
        }
        
        for (Integer p : path) {
            this.father.put(p, point);
        }
        return point;
    }

    /*
     * @param a: An integer
     * @param b: An integer
     * @return: A boolean
     */
    public boolean query(int a, int b) {
        // write your code here
        
        int root_a = this.find(a);
        int root_b = this.find(b);
        
        return root_a == root_b;
        
        
    }

    public static void main(String[] args) {

        ConnectingGraph cg = new ConnectingGraph(8);
        ArrayList<Boolean> results = new ArrayList<>();

        results.add(cg.query(1, 2));

        cg.connect(1, 2);

        results.add(cg.query(1, 3)); 

        cg.connect(2, 4);

        results.add(cg.query(1, 4));

        System.out.println(results);
    }
}